akka-actor-typed/src/main/scala/akka/actor/typed/ActorSystem.scala

```diff
@@ -221,13 +221,43 @@ object ActorSystem {
     val appConfig = config.getOrElse(ConfigFactory.load(cl))
     val setup = ActorSystemSetup(BootstrapSetup(classLoader, config, executionContext))
     val untyped = new a.ActorSystemImpl(name, appConfig, cl, executionContext,
-      Some(PropsAdapter(() ⇒ guardianBehavior, guardianProps)), setup)
+      Some(PropsAdapter(() ⇒ delayedGuardianStart(guardianBehavior), guardianProps)), setup)
     untyped.start()
 
     val adapter: ActorSystemAdapter.AdapterExtension = ActorSystemAdapter.AdapterExtension(untyped)
+    untyped.guardian ! Started
     adapter.adapter
   }
 
+  private[akka] case object Started
+  private def delayedGuardianStart[T](guardianBehavior: Behavior[T], buffer: List[Any] = Nil): Behavior[Any] =
+    scaladsl.Actor.immutable[Any] { (sctx, msg) ⇒
+      // delay actual initialization until the actor system is started as the actor may touch
+      // other parts of the actor system which would not be ready unless delayed
+      val ctx = sctx.asInstanceOf[ActorContext[T]]
+      msg match {
+        case Started ⇒
+          val endState = buffer.reverseIterator.foldLeft(Behavior.undefer(guardianBehavior, ctx)) { (behavior, buffered) ⇒
```

Comment:
isn't `undefer` needed for each returned behavior, i.e. also inside the fold?

Reply:
Yes, good catch.

Reply:
there are more bugs here I think, e.g. when `same` is returned, probably wait with PR until after my Stashing PR and use that instead